src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/ResourceList.java

```diff
@@ -0,0 +1,136 @@
+ * questions.
+ */
+
+package jdk.internal.foreign;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.VarHandle;
+
+public abstract class ResourceList implements Runnable {
+    ResourceCleanup fst;
+
+    abstract void add(ResourceCleanup cleanup);
+
+    abstract void cleanup();
+
+    public final void run() {
+        cleanup(); // cleaner interop
+    }
+
+    final void cleanup(ResourceCleanup first) {
+        ResourceCleanup current = first;
+        while (current != null) {
+            current.cleanup();
+            current = current.next;
+        }
+    }
+
+    public static abstract class ResourceCleanup {
+        ResourceCleanup next;
+
+        public abstract void cleanup();
+
+        final static ResourceCleanup DUMMY_CLEANUP = new ResourceCleanup() {
+            @Override
+            public void cleanup() {
+                // do nothing
+            }
+        };
+
+        static ResourceCleanup ofRunnable(Runnable cleanupAction) {
+            return new ResourceCleanup() {
+                @Override
+                public void cleanup() {
+                    cleanupAction.run();
+                }
+            };
+        }
+    }
+
+    static class ConfinedResourceList extends ResourceList {
+        @Override
+        void add(ResourceCleanup cleanup) {
+            if (fst != ResourceCleanup.DUMMY_CLEANUP) {
+                cleanup.next = fst;
+                fst = cleanup;
+            } else {
+                throw new IllegalStateException("Already closed!");
+            }
+        }
+
+        @Override
+        void cleanup() {
+            if (fst != ResourceCleanup.DUMMY_CLEANUP) {
+                ResourceCleanup prev = fst;
+                fst = ResourceCleanup.DUMMY_CLEANUP;
+                cleanup(prev);
+            }
+        }
+    }
+
+    static class SharedResourceList extends ResourceList {
+
+        static final VarHandle FST;
+
+        static {
+            try {
+                FST = MethodHandles.lookup().findVarHandle(ResourceList.class, "fst", ResourceCleanup.class);
+            } catch (Throwable ex) {
+                throw new ExceptionInInitializerError();
+            }
+        }
+
+        @Override
+        void add(ResourceCleanup cleanup) {
+            while (true) {
+                ResourceCleanup prev = (ResourceCleanup) FST.getAcquire(this);
+                cleanup.next = prev;
+                ResourceCleanup newSegment = (ResourceCleanup) FST.compareAndExchangeRelease(this, prev, cleanup);
+                if (newSegment == ResourceCleanup.DUMMY_CLEANUP) {
+                    // too late
+                    throw new IllegalStateException("Already closed");
+                } else if (newSegment == prev) {
+                    return; //victory
+                }
+                // keep trying
+            }
+        }
+
+        void cleanup() {
+            if (fst != ResourceCleanup.DUMMY_CLEANUP) {
```

Comment:
Probably - but, to be clear - at this point (and I need to add doc on this), we are only interested about add vs. close races - not close vs. close. So, the only "bad" thing that could happen is that some other thread adds to this list while we're closing it. So I guess in this case I omitted the read acquire because the only case where you could read a DUMMY_CLEANUP is if somebody has already stored that in there (otherwise you'll see something else, and move on). But no harm in making it more robust.

Another way we could go would be to manage races in the MemoryScope impl. That is, in a shared memory scope we could have some intermediate state (e.g. ADDING) which we go through and which is mutually exclusive w.r.t. CLOSE. I didn't want to go down that path since then in ResourceList I'd have to deal with add vs. add races anyway.

Reply:
Ok, that makes sense. I don't think an ADDING state is needed, but would be nice if you could comment this part.